File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res(nums1.size(), -1);
        stack<int> s;
        map<int, int> nums1Idx;
        for (size_t i = 0; i < nums1.size(); ++i) nums1Idx[nums1[i]] = i;

        for (int i = 0; i < nums2.size(); ++i) {
            int curr = nums2[i];
            while ((!s.empty()) && curr > s.top()) {
                res[nums1Idx[s.top()]] = curr;
                s.pop();
            };
            if (nums1Idx.find(curr) != nums1Idx.end()) s.push(curr);
        };
        return res;
    };
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:

        res, mapping = [], {}
        maxEl = nums2[-1]

        for i in range(len(nums2)-1, -1, -1):
            if nums2[i] >= maxEl:
                maxEl = nums2[i]
                mapping[nums2[i]] = -1

            else:
                j = i
                while j < len(nums2) and nums2[j] <= nums2[i]:
                    j += 1
                if j == len(nums2):
                    mapping[nums2[i]] = -1
                else:
                    mapping[nums2[i]] = nums2[j] 

        for el in nums1:
            res.append(mapping[el])

        return res